opentelemetry-otlp/src/exporter/http/mod.rs

```diff
@@ -289,59 +338,75 @@ impl OtlpHttpClient {
     fn build_trace_export_body(
         &self,
         spans: Vec<SpanData>,
-    ) -> Result<(Vec<u8>, &'static str), String> {
+    ) -> Result<(Vec<u8>, &'static str, Option<&'static str>), String> {
         use opentelemetry_proto::tonic::collector::trace::v1::ExportTraceServiceRequest;
         let resource_spans = group_spans_by_resource_and_scope(spans, &self.resource);
 
         let req = ExportTraceServiceRequest { resource_spans };
-        match self.protocol {
+        let (body, content_type) = match self.protocol {
             #[cfg(feature = "http-json")]
             Protocol::HttpJson => match serde_json::to_string_pretty(&req) {
-                Ok(json) => Ok((json.into_bytes(), "application/json")),
-                Err(e) => Err(e.to_string()),
+                Ok(json) => (json.into_bytes(), "application/json"),
+                Err(e) => return Err(e.to_string()),
             },
-            _ => Ok((req.encode_to_vec(), "application/x-protobuf")),
-        }
+            _ => (req.encode_to_vec(), "application/x-protobuf"),
+        };
+
+        let (compressed_body, content_encoding) = self.compress_body(body)?;
+        Ok((compressed_body, content_type, content_encoding))
```

Comment:
nit: The variable gives the wrong impression that compressed_body is always compressed. But if Compression is None, then its just returning existing body as-is.
May be:
if (self.compression is None)
{
 Ok(body,content_type,None)
}
else
{
call self.compress_body(...),
}

Reply:
I've renamed it to "process_body"; this way we don't have confusion about whether or not the caller needs to also check if the body needs compressing, we just always pass the body through the call!